### DISCUSSION TITLE
Jest testing extensions: mocking internal ItemsService

### DISCUSSION THREAD

**Discussion by keesvanbemmel** [directus/directus#8884] [Archive]
Hi!

While developing extensions, it's obvious I should mock calls to the ItemsService and others. Let's say my function does this:
```
async function middlewareFunction(req, res, next, context) {
    const { ItemsService } = context.services;
    const { schema } = req;
    const postsService = new ItemsService('posts', {
      schema,
    });

    const currentPost = await postsService.readOne('1', {
      fields: ['*'],
    });
}
```

Are there any best practices or examples on how to mock the Request, Response, context and the ItemsService.readOne?

Any help is welcome!

Kees.

**Comment by benhaynes** [comment]
CC @jaycammarano

**Comment by rijkvanzanten** [comment]
Most services are implementing the same shared abstract interface, so that could be a good place to find the results you should mock:

https://github.com/directus/directus/blob/main/api/src/types/services.ts
